File: android/app/src/main/java/com/rate/rgbwallet/RGBNodeModule.kt
package com.rate.rgbwallet

import android.content.Context
import android.content.Intent
import android.os.Build
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class RGBNodeModule : Module() {
    private var isServiceRunning = false
    private val binaryName = "rgb-lightning-node"

    override fun definition() = ModuleDefinition {
        Name("RGBNode")

        Function("startNode") { options: Map<String, Any> ->
            try {
                // Extract binary if needed
                extractBinaryIfNeeded()

                // Start the service
                val intent = Intent(context, RGBNodeService::class.java).apply {
                    options.forEach { (key, value) ->
                        when (value) {
                            is String -> putExtra(key, value)
                            is Int -> putExtra(key, value)
                            is Boolean -> putExtra(key, value)
                        }
                    }
                }

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    context.startForegroundService(intent)
                } else {
                    context.startService(intent)
                }

                isServiceRunning = true

                mapOf(
                    "success" to true,
                    "port" to (options["daemonPort"] ?: 3000),
                    "lightningPort" to (options["lightningPort"] ?: 9735)
                )
            } catch (e: Exception) {
                throw Exception("Failed to start RGB node: ${e.message}")
            }
        }

        Function("stopNode") {
            try {
                context.stopService(Intent(context, RGBNodeService::class.java))
                isServiceRunning = false
                mapOf("success" to true)
            } catch (e: Exception) {
                throw Exception("Failed to stop RGB node: ${e.message}")
            }
        }

        Function("isNodeRunning") {
            mapOf("isRunning" to isServiceRunning)
        }
    }

    private fun extractBinaryIfNeeded() {
        val binaryFile = File(context.filesDir, binaryName)

        if (binaryFile.exists()) {
            return
        }

        GlobalScope.launch(Dispatchers.IO) {
            withContext(Dispatchers.IO) {
                context.assets.open(binaryName).use { input ->
                    FileOutputStream(binaryFile).use { output ->
                        input.copyTo(output)
                    }
                }

                // Set executable permissions
                binaryFile.setExecutable(true)
            }
        }
    }
} 

File: ios/RGBNodeModule.swift
import Foundation
import ExpoModulesCore

public class RGBNodeModule: Module {
    public func definition() -> ModuleDefinition {
        Name("RGBNode")
        
        Function("startNode") { (options: [String: Any]) -> Promise in
            Promise { promise in
                do {
                    // Extract binary if needed
                    try self.extractBinaryIfNeeded()
                    
                    // Start the node process
                    try self.startNodeProcess(options: options)
                    
                    promise.resolve([
                        "success": true,
                        "daemon_port": options["daemon_listening_port"] ?? 3000,
                        "ldk_peer_port": options["ldk_peer_listening_port"] ?? 9735
                    ])
                } catch {
                    promise.reject("ERR_NODE_START", "Failed to start RGB Lightning Node: \(error.localizedDescription)")
                }
            }
        }
        
        Function("stopNode") { () -> Promise in
            Promise { promise in
                do {
                    try self.stopNodeProcess()
                    promise.resolve(["success": true])
                } catch {
                    promise.reject("ERR_NODE_STOP", "Failed to stop RGB Lightning Node: \(error.localizedDescription)")
                }
            }
        }
        
        Function("isNodeRunning") { () -> Promise in
            Promise { promise in
                let isRunning = self.checkNodeProcess()
                promise.resolve(["isRunning": isRunning])
            }
        }
    }
    
    private var nodeProcess: Process?
    private let binaryName = "rgb-lightning-node"
    
    private func extractBinaryIfNeeded() throws {
        let fileManager = FileManager.default
        let documentsPath = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let binaryPath = documentsPath.appendingPathComponent(binaryName)
        
        // Check if binary already exists
        if fileManager.fileExists(atPath: binaryPath.path) {
            return
        }
        
        // Get binary from bundle
        guard let bundleBinaryPath = Bundle.main.path(forResource: binaryName, ofType: nil) else {
            throw NSError(domain: "com.rate.rgbwallet", code: -1, userInfo: [
                NSLocalizedDescriptionKey: "RGB Lightning Node binary not found in bundle"
            ])
        }
        
        // Copy binary to documents directory
        try fileManager.copyItem(atPath: bundleBinaryPath, toPath: binaryPath.path)
        
        // Set executable permissions
        try fileManager.setAttributes([.posixPermissions: 0o755], ofItemAtPath: binaryPath.path)
    }
    
    private func startNodeProcess(options: [String: Any]) throws {
        let fileManager = FileManager.default
        let documentsPath = fileManager.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let binaryPath = documentsPath.appendingPathComponent(binaryName)
        let dataDir = documentsPath.appendingPathComponent("rgb-data")
        
        // Create data directory if it doesn't exist
        try fileManager.createDirectory(at: dataDir, withIntermediateDirectories: true)
        
        // Create and configure process
        let process = Process()
        process.executableURL = binaryPath
        
        // Configure arguments based on options
        var arguments: [String] = []
        arguments.append(dataDir.path)
        
        if let daemonPort = options["daemon_listening_port"] as? Int {
            arguments.append("--daemon-listening-port")
            arguments.append("\(daemonPort)")
        }
        
        if let ldkPort = options["ldk_peer_listening_port"] as? Int {
            arguments.append("--ldk-peer-listening-port")
            arguments.append("\(ldkPort)")
        }
        
        if let network = options["network"] as? String {
            arguments.append("--network")
            arguments.append(network)
        }
        
        process.arguments = arguments
        
        // Setup pipes for output
        let outputPipe = Pipe()
        let errorPipe = Pipe()
        process.standardOutput = outputPipe
        process.standardError = errorPipe
        
        // Start process
        try process.run()
        self.nodeProcess = process
        
        // Monitor process in background
        DispatchQueue.global(qos: .background).async {
            process.waitUntilExit()
            
            // Handle process termination
            if process.terminationStatus != 0 {
                let errorData = errorPipe.fileHandleForReading.readDataToEndOfFile()
                let errorOutput = String(data: errorData, encoding: .utf8)
                print("RGB Lightning Node process terminated with error: \(errorOutput ?? "Unknown error")")
            }
        }
    }
    
    private func stopNodeProcess() throws {
        guard let process = self.nodeProcess else { return }
        
        process.terminate()
        self.nodeProcess = nil
    }
    
    private func checkNodeProcess() -> Bool {
        guard let process = self.nodeProcess else { return false }
        return process.isRunning
    }
} 